src/ripple/overlay/MessageScheduler.h

```diff
@@ -0,0 +1,496 @@
+#ifndef RIPPLE_OVERLAY_MESSAGESCHEDULER_H_INCLUDED
+#define RIPPLE_OVERLAY_MESSAGESCHEDULER_H_INCLUDED
+
+#include <ripple/basics/chrono.h>
+#include <ripple/basics/random.h>
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/overlay/Peer.h>
+#include <ripple/overlay/impl/ProtocolMessage.h>
+#include <ripple.pb.h>
+
+#include <boost/asio/basic_waitable_timer.hpp>
+#include <boost/asio/io_service.hpp>
+#include <google/protobuf/message.h>
+
+#include <atomic>
+#include <cassert>
+#include <chrono>
+#include <cstdint>
+#include <functional>
+#include <iterator>
+#include <memory>
+#include <mutex>
+#include <ostream>
+#include <random>
+
+namespace ripple {
+
+/**
+ * We must hold idle peers by `std::weak_ptr` so that they can destruct upon
+ * disconnect.
+ * When we remove channels in `remove`, we need to find them by peer ID,
+ * because `std::weak_ptr` is not equality-comparable with anything,
+ * but we do not want to lock the `std::weak_ptr` to get that ID,
+ * so we copy it.
+ */
+struct Channel
+{
+    Peer::id_t peerId;
+    std::weak_ptr<Peer> peer;
+};
+
+inline Channel
+toChannel(std::shared_ptr<Peer> peer)
+{
+    return {peer->id(), peer};
+}
+
+class MessageScheduler
+{
+public:
+    enum class FailureCode {
+        // The request timed out.
+        TIMEOUT,
+        // The peer disconnected.
+        DISCONNECT,
+        // The message scheduler is shutting down.
+        SHUTDOWN,
+    };
+
+    using RequestId = uint32_t;
```

Comment:
What about associated types in containers like `reference`, `const_reference`, `size_type`, etc.? Many of these are simple types that can be written out by hand by users, but isn't it better to have a name for that type expression?

I want these aliases so that there is one place to change the type if needed, and so that parameter types are readable, especially in the absence of a parameter name. I don't expect all distinct type names to be inconvertible.

Reply:
You have an excellent point.  I consider the `reference`... typedefs to be a bit safer because they usually have non-builtin underlying types.  So they usually _are_ nice safe types (or pointers or references to them) you have in your hand.

`size_type` is a closer analogy, since it almost always aliases some kind of builtin integer.  And, yeah, I've been trained to have the right expectations of `size_type`.  Maybe I can be trained to have the same reaction to `RequestId`.

Really, I'm just expressing a personal opinion here.  I've been tripped up by this kind of alias for builtin types, particularly integers, in the past.  Maybe that's just because I'm sloppy.  If you think the alias is important you should keep it.